datafusion/physical-plan/src/dynamic_filters.rs

```diff
@@ -0,0 +1,320 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::{
+    any::Any,
+    hash::Hash,
+    sync::{Arc, RwLock},
+};
+
+use datafusion_common::{
+    tree_node::{Transformed, TransformedResult, TreeNode},
+    Result,
+};
+use datafusion_expr::ColumnarValue;
+use datafusion_physical_expr::{expressions::lit, utils::conjunction, PhysicalExpr};
+
+/// A source of dynamic runtime filters.
+///
+/// During query execution, operators implementing this trait can provide
+/// filter expressions that other operators can use to dynamically prune data.
+///
+/// See `TopKDynamicFilterSource` in datafusion/physical-plan/src/topk/mod.rs for examples.
+pub trait DynamicFilterSource: Send + Sync + std::fmt::Debug + 'static {
+    /// Take a snapshot of the current state of filtering, returning a non-dynamic PhysicalExpr.
+    /// This is used to e.g. serialize dynamic filters across the wire or to pass them into systems
+    /// that won't use the `PhysicalExpr` API (e.g. matching on the concrete types of the expressions like `PruningPredicate` does).
+    /// For example, it is expected that this returns a relatively simple expression such as `col1 > 5` for a TopK operator or
+    /// `col2 IN (1, 2, ... N)` for a HashJoin operator.
+    fn snapshot_current_filters(&self) -> Result<Vec<Arc<dyn PhysicalExpr>>>;
+}
+
+#[derive(Debug)]
+pub struct DynamicFilterPhysicalExpr {
```

Comment:
Could you add in some small context about what a `DynamicFilterPhysicalExpr` is?